src/instructions/segmentation.rs

```diff
@@ -1,210 +1,270 @@
-/// is a valid extra segment descriptor.
-#[inline]
-pub unsafe fn load_es(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov es, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+macro_rules! segment64_impl {
+    ($type:ty, $name:literal, $asm_rd:ident, $asm_wr:ident) => {
+        impl Segment64 for $type {
+            unsafe fn read_base() -> VirtAddr {
+                #[cfg(feature = "inline_asm")]
+                {
+                    let val: u64;
+                    asm!(concat!("rd", $name, "base {}"), out(reg) val, options(nomem, nostack, preserves_flags));
+                    VirtAddr::new_unsafe(val)
+                }
+                #[cfg(not(feature = "inline_asm"))]
+                VirtAddr::new_unsafe(crate::asm::$asm_rd())
+            }
+
+            unsafe fn write_base(base: VirtAddr) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("wr", $name, "base {}"), in(reg) base.as_u64(), options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_es(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_wr(base.as_u64());
+            }
+        }
+    };
 }
 
-/// Reload fs segment register.
+/// Code Segment
 ///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid fs segment descriptor.
-#[inline]
-pub unsafe fn load_fs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov fs, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// The segment base and limit are unused in 64-bit mode. Only the L (long), D
+/// (default operation size), and DPL (descriptor privilege-level) fields of the
+/// descriptor are recognized. So chaning the segment register can be used to
+/// change privilege level or enable/disable long mode.
+#[derive(Debug)]
+pub struct CS;
+impl Segment for CS {
+    get_reg_impl!("cs", x86_64_asm_get_cs);
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_fs(sel.0);
+    unsafe fn set_reg(sel: SegmentSelector) {
+        // Note this is special since we cannot directly move to cs. Instead we
+        // push the new segment selector and return value on the stack and use
+        // retfq to reload cs and continue at 1:.
+        #[cfg(feature = "inline_asm")]
+        asm!(
+            "push {sel}",
+            "lea {tmp}, [1f + rip]",
+            "push {tmp}",
+            "retfq",
+            "1:",
+            sel = in(reg) u64::from(sel.0),
+            tmp = lateout(reg) _,
+            options(preserves_flags),
+        );
+
+        #[cfg(not(feature = "inline_asm"))]
+        crate::asm::x86_64_asm_set_cs(u64::from(sel.0));
+    }
 }
 
-/// Reload gs segment register.
+/// Stack Segment
 ///
-/// ## Safety
+/// Entirely unused in 64-bit mode, setting the segment register does nothing.
```

Comment:
While the SS register is entirely useless for 64-bit long-mode, it is often still set because of syscall/sysret. Should we mention that?

Reply:
Yes, I think that would be good to mention.

Reply:
Added in https://github.com/rust-osdev/x86_64/pull/258/commits/9c4a2c901a0787ef4368d4f21c13cf3c970f066d